README.md
# RUSH-or-NOT

A small project that record you RUSH or NOT

## LEVEL of RUSHER

![image-20220227235427258](http://tallestdaisy.oss-cn-beijing.aliyuncs.com/img/image-20220227235427258.png)

## DEMO DISPLAY

![image-20220227235413376](http://tallestdaisy.oss-cn-beijing.aliyuncs.com/img/image-20220227235413376.png)

## 需求分析

- 能够保存个人的**rush记录**
  - 存储个人历史rush记录
    - 定义第i个rush的生命周期为从第i次rush开始到第i+1次结束
    - 存储每一个已经结束的rush
    - 每一个rush的信息包括
      - rush编号
      - rush开始时间
      - rush结束时间
      - 本次持续时间
      - 等级
- 能够显示**个人的rush信息**
  - 界面展示rush信息
    - 当前时间
    - 上次rush时间
    - 本次持续时间
    - 下一等级时间
    - 当前等级
    - 下个等级
    - 历史最长持续时间
    - 历史最高等级
- 能够通过**GUI交互**，记录rush
  - 点击rush按钮代表rush
    - 点击rush按钮，更新历史记录，也更新界面显示信息

## 系统设计

将系统分为以下模块（目前阶段不考虑系统架构，仅仅考虑功能模块）：

- 文件存储模块
  - 对文件的所有操作，抽象成方法，根据此模块提供的方法对文件进行操作
  - 每一种操作都抽象成一个方法
    - 读出current.txt存储的上次坠机时间？**冗余**
    - 修改current.txt的上次坠机时间？**冗余**
    - 读出history.txt中的所有历史坠机记录
    - 读出history.txt中的上次历史坠机记录
    - 读出history.txt中的上次历史坠机**时间**
    - 增加history.txt内容，在尾部增加一条坠机记录
- 界面展示模块
  - 界面中所有显示的元素抽象成方法，main使用此模块的方法产生各种界面。
  - 每一个界面显示元素都抽象成一个方法
    - 上次坠机百京时间，标题+动态数据
    - 道友当前百京时间，标题+动态数据
    - 道友当前等级，标题+动态数据
    - 道友下一割等级，标题+动态数据
    - 美甲的窃魂卷持续时长，标题+动态数据
    - 突破倒计时，标题+动态数据
- 人机交互模块
  - 人机交互的操作都抽象成方法，人机交互的时候触发此模块的方法。
  - 每一个人机交互事件，都抽象成一个方法。
    - 坠机
      - 修改数据，记录当前信息，写入到历史坠机记录中
      - 界面显示，将界面显示的动态元素更新。





resourse/history.txt
1 1645874396.3903117 1605874396.3903117 40000000 心渡天劫
2 1645874396.3903117 1605874396.3903117 40000000 心渡天劫
3 1645874396.3903117 1605874396.3903117 40000000 心渡天劫
4 1645874396.3903117 1705874396.3903117 40000000 心渡天劫
4 1645874396.3903117 1705874396.3903117 60000000.0 痴傻无知


test.py


import os

from matplotlib.pyplot import close
from numpy import record


cwd = os.getcwd()
history_path = cwd+"\\resourse\\history.txt"
#读出history.txt中的所有历史坠机记录
def read_history_total_records(path):
    f = open(path,'r',encoding="utf-8")
    record_list = []
    for line in f.readlines():
        record_list.append(line.split(' '))
    f.close()
    return record_list

#读出history.txt中的上次历史坠机记录
def read_history_last_record(path):
    return read_history_total_records(path)[-1]

#读出history.txt中的上次历史坠机时间
def read_history_last_datetime(path):
    return read_history_last_record(path)[2]

#增加history.txt内容，在尾部增加一条坠机记录
def append_history_a_new_record(path, start_time = 1645874396.3903117, end_time = 1705874396.3903117, level_name = '痴傻无知'):
    last_time = end_time - start_time
    id = len(read_history_total_records(path))
    new_line = str(id) + ' ' + str(start_time) + ' ' + str(end_time) + ' ' + str(last_time) + ' ' + level_name + '\n'
    f = open(path,'a',encoding="utf-8")
    f.write(new_line)
    f.close()

record_list = read_history_total_records(history_path)
# for line in record_list:
#     print(line)

last_record = read_history_last_record(history_path)
# print(last_record)

last_record_datetime = read_history_last_datetime(history_path)
# print(last_record_datetime)

append_history_a_new_record(history_path)

initLevel.py

from sys import dllhandle


f1 = open("resourse\\level_raw.txt","r",encoding="utf-8")
f2 = open("resourse\\level.txt",'w',encoding='utf-8')
level_list = []
current_word=f1.read(4)
while current_word:
    level_list.append(current_word)
    current_word=f1.read(4)
# print(level_list)
day_list = []
for i in range(1,125,3):
    day_list.append(i)
for i in range(128,193,4):
    day_list.append(i)
for i in range(197,308,5):
    day_list.append(i)
for i in range(312,469,6):
    day_list.append(i)
for i in range(475,518,7):
    day_list.append(i)
for i in range(525,614,8):
    day_list.append(i)
day_list.append(667)
day_list.append(778)
# print(day_list)
# print(len(level_list))
# print(len(day_list))
for i in range(len(level_list)):
    f2.writelines(level_list[i]+" "+str(day_list[i]) + '\n')
f1.close()
f2.close()


t.py
import tkinter as tk
import time
import os
import locale

from matplotlib.pyplot import text
locale.setlocale(locale.LC_CTYPE, 'chinese')

#local parameter
WIDTH = 35
HEIGHT = 2
root = tk.Tk()
root.title('修仙辅助器')
root.geometry('500x500')
# 上次rush的时间
last_rush_datetime = -1
# 当前时间
now_datetime = -1
# 本次修炼时间
current_last_time = -1
# 到下一个等级剩余时间
next_level_remain_time = -1
# 当前等级
current_level = -1
# 下一个等级
next_level = ""
# 历史rush记录
history_datetime_list = []
# 所有等级列表
level_list = []

#创建界面元素
#上次坠机时间
last_rush_datetime_static_label = tk.Label(root, text='上次坠机时间',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
last_rush_datetime_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
#道友当前时间
now_datetime_static_label = tk.Label(root, text='道友当前时间',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
now_datetime_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
#道友当前等级
current_level_static_label = tk.Label(root, text='道友当前等级',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
current_level_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
#道友下一割等级
next_level_static_label = tk.Label(root, text='道友下一割等级',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
next_level_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
#美甲的窃魂卷持续时长
current_last_time_static_label = tk.Label(root, text='美甲的窃魂卷持续时长',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
current_last_time_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
#突破倒计时
next_level_remain_time_static_label = tk.Label(root, text='突破倒计时',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)
next_level_remain_time_dynamic_label = tk.Label(root, text='NULL',anchor="e", fg='blue', font=("黑体", 12),width=WIDTH,height=HEIGHT)

#为桌面元素布局
def grid_all():
    last_rush_datetime_static_label.grid(row=1,column=1)
    last_rush_datetime_dynamic_label.grid(row=1,column=2)
    #道友当前时间
    now_datetime_static_label.grid(row=2,column=1)
    now_datetime_dynamic_label.grid(row=2,column=2)
    #道友当前等级
    current_level_static_label.grid(row=3,column=1)
    current_level_dynamic_label.grid(row=3,column=2)
    #道友下一割等级
    next_level_static_label.grid(row=4,column=1)
    next_level_dynamic_label.grid(row=4,column=2)
    #美甲的窃魂卷持续时长
    current_last_time_static_label.grid(row=5,column=1)
    current_last_time_dynamic_label.grid(row=5,column=2)
    #突破倒计时
    next_level_remain_time_static_label.grid(row=6,column=1)
    next_level_remain_time_dynamic_label.grid(row=6,column=2)

#初始化
def init():
    #最开始要布局
    grid_all()
    global last_rush_datetime
    global now_datetime
    global level_list
    global current_level
    #初始化当前时间
    now_datetime = time.time()
    cwd = os.getcwd()
    f1 = open(cwd+"\\resourse\\current.txt","r",encoding="utf-8")
    f2 = open(cwd+"\\resourse\\level.txt","r",encoding="utf-8")
    #初始化上次rush时间
    last_rush_datetime = int(f1.readline().split('.')[0])
    #初始化level等级列表
    for line in f2.readlines():
        level_name , level_days = line.replace("\n",'').split(' ')
        level_list.append([level_name,int(level_days)])
    #初始化当前level
    for i in range(len(level_list)):
        # print(now_datetime-last_rush_datetime)
        if(now_datetime-last_rush_datetime  < (level_list[i+1][1])*86400):
            current_level = i
            break
    #初始化动态label的值
    last_rush_datetime_dynamic_label.configure(text=time.strftime("%Y年-%m月-%d号 %H小时%M分%S秒",last_rush_datetime))
    #tudo 上次的中断位置
    # current_level_dynamic_label.configure(text=)
    # last_rush_datetime_dynamic_label.configure(text=time.strftime("%Y年-%m月-%d号 %H小时%M分%S秒",last_rush_datetime))
    # last_rush_datetime_dynamic_label.configure(text=time.strftime("%Y年-%m月-%d号 %H小时%M分%S秒",last_rush_datetime))

#实时刷新
#Todo！
def update():
    global now_datetime
    global current_last_time
    global next_level_remain_time
    global current_level
    #更新datetime，time数据
    #更新实时数据
    now_datetime = time.time()
    current_last_time = now_datetime - last_rush_datetime
    next_level_remain_time = level_list[current_level+1][1]*86400 - current_last_time

    #实时判断是否升级
    #升级
    if next_level_remain_time <=0:
        current_level += 1
        #更新界面显示
        current_level_dynamic_label_str = f"Level {current_level} {level_list[current_level][0]} 嘻嘻"
        current_level_dynamic_label.configure(text=current_level_dynamic_label_str)
        next_level_remain_time_dynamic_label_str = f"Level {current_level+1} {level_list[current_level+1][0]} 嘻嘻嘻"
        next_level_remain_time_dynamic_label.configure(text=next_level_remain_time_dynamic_label_str)

    # 获取时间并转为字符串
    now_datetime_dynamic_label_str = time.strftime("%Y年-%m月-%d号 %H小时%M分%S秒",time.localtime(now_datetime))  
    current_last_time_dynamic_label_str = f"{int((current_last_time)/86400)}天"+time.strftime("%H小时%M分%S秒",time.gmtime(current_last_time))
    next_level_remain_time_dynamic_label_str = f"{int(next_level_remain_time/86400)}天"+time.strftime("%H小时%M分%S秒",time.gmtime(next_level_remain_time%86400))
    # 重新设置标签文本
    now_datetime_dynamic_label.configure(text=now_datetime_dynamic_label_str)
    current_last_time_dynamic_label.configure(text=current_last_time_dynamic_label_str)
    next_level_remain_time_dynamic_label.configure(text=next_level_remain_time_dynamic_label_str)
    # 每隔一秒调用函数gettime自身获取时间
    root.after(1000, update)

cwd = os.getcwd()
history_path = cwd+"\\resourse\\history.txt"
#读出history.txt中的所有历史坠机记录
def read_history_total_records(path):
    f = open(path,'r',encoding="utf-8")
    record_list = []
    for line in f.readlines():
        record_list.append(line.split(' '))
    f.close()
    return record_list

#读出history.txt中的上次历史坠机记录
def read_history_last_record(path):
    return read_history_total_records(path)[-1]

#读出history.txt中的上次历史坠机时间
def read_history_last_datetime(path):
    return read_history_last_record(path)[2]

#增加history.txt内容，在尾部增加一条坠机记录
def append_history_a_new_record(path, start_time = 1645874396.3903117, end_time = 1705874396.3903117, level_name = '痴傻无知'):
    last_time = end_time - start_time
    id = len(read_history_total_records(path))
    new_line = str(id) + ' ' + str(start_time) + ' ' + str(end_time) + ' ' + str(last_time) + ' ' + level_name + '\n'
    f = open(path,'a',encoding="utf-8")
    f.write(new_line)
    f.close()


# var = tk.StringVar()    # 这时文字变量储存器
# l = tk.Label(root, 
#     textvariable=var,    # 标签的文字
#     bg='green',     # 标签背景颜色
#     font=('Arial', 12),     # 字体和字体大小
#     width=10, height=2  # 标签长宽(以字符长度计算)
#     )
# l.grid(row=0,column=0,pady=2)

# on_hit=False
 
# #按钮的函数
# def hit_me():
#     global on_hit#声明全局变量
#     if on_hit==False:
#         on_hit=True
#         var.set('You hit me!')
#     else:
#         on_hit=False
#         var.set('')
# #按钮
# b=tk.Button(root,text='点我',width=10,height=2,command=hit_me)#点击按钮执行一个名为“hit_me”的函数
# b.grid(row=0,column=2,pady=2)

init()

update()


root.mainloop()
